User can anonymize an opinion

chadbrower0 opened the pull request:
This change adds buttons in the proposal metadata, to edit / anonymize / delete the participant's opinion.  Anonymizing the opinion also anonymizes the partcipant's reasons for the proposal.  

<img width="499" alt="image" src="https://github.com/Considerit/ConsiderIt/assets/17461871/2d3a08f4-91b5-4ff3-b605-e8d6e926c728">

* The anonymous-mask-icon is a hand-edited SVG.  I tried to keep it simple, for clarity at small size.  But it could be better.
* Maybe opinion-anonymization should also apply to the participant's comments on the proposal.

tkriplean commented:
Oh my goodness, this is awesome. Thanks so much! I like the mask svg work you did, I think it is fine. Here are some notes on getting it to the finish line:

- Opinions aren't updating in the histogram after a slidergram drag. If you drag, and then refresh, the opinion shows up in the right place. It works the first time you add your opinion, but not subsequent drags. Oh! Strangely enough, it works if your opinion is anonymous.

- "This will remove your opinion from the histogram, but keep the pros and cons and comments that you have written for this proposal. Are you sure you wish to remove?" This is wrong: in order to post a pro or con point, you need to have a published opinion. So when you delete your opinion, the pros and cons you wrote will be deleted. (Note, I don't think this is the best way to handle it. I think it would be best if it worked like the following: your pros and cons are deleted when your opinion is deleted, if no one else has included them. However, for pros and cons that other people have used, the author of the respective points is re-assigned to a universal "[deleted]" (or something similar) account. (A similar set of problems happens in the rare case that someone deletes their account, which should be handled more gracefully -- comments disappear mid-thread, and pros/cons are deleted regardless of how many people included them.))

- When someone has not added their opinion to a proposal yet, I think the area should just read "Add your opinion", instead of "Your opinion: [edit icon]". I think this is better overall, but also fixes the problem where you end up with ["Your opinion:"] on small screens (because of the nice work you did removing the edit icon when the screen width gets small).

- It would be nice to give some indicator that your opinion is anonymous in the opinion crafting view. Maybe next to the "Remove your opinion" button below the dashed your opinion box. I'll add a mockup.

tkriplean commented:
<img width="726" alt="opinion-crafting-anonymous" src="https://github.com/Considerit/ConsiderIt/assets/707039/360b51d8-5f5d-4e82-97cc-67ce20a7b227">
<img width="728" alt="opinion-crafting-anonymized" src="https://github.com/Considerit/ConsiderIt/assets/707039/4e93a66e-03e3-433e-8d2a-5cf59a351317">

chadbrower0 commented:
> Opinions aren't updating in the histogram after a slidergram drag... it works if your opinion is anonymous.

Fixed in commit [403f9ec](https://github.com/Considerit/ConsiderIt/pull/183/commits/403f9ec4d4019449dd8f9d08d2fed181bf016673)

I will continue working on the other modifications.

chadbrower0 commented:
> when you delete your opinion, the pros and cons you wrote will be deleted.

Updated prompts to more accurately describe deletion.

> "Add your opinion", instead of "Your opinion: [edit icon]".

Reused the existing getOpinionPrompt() which supplies "Give your opinion".  We could change that function to yield "Add your opinion" in all its usages, though I suggest doing that in a separate change that considers the variety of usages of that function.

>  in the opinion crafting view... "Remove your opinion" button below the dashed your opinion box. I'll add a mockup.

Implemented per mockup

> I'm not sure I like the trash can over a more subtle x

Done

tkriplean commented:
Well, this pull request has a lot of pork in it now, ha. Major changes described at https://traviskriplean.com/lively-anonymity-overhaul-for--bdh9kk.
